core/common/src/main/java/alluxio/collections/UnmodifiableArrayList.java

```diff
@@ -0,0 +1,311 @@
+  }
+
+  @Override
+  public boolean remove(Object o) {
+    throw new UnsupportedOperationException(
+        "modification is not supported in UnmodifiableArrayList");
+  }
+
+  @Override
+  public T remove(int index) {
+    throw new UnsupportedOperationException(
+        "modification is not supported in UnmodifiableArrayList");
+  }
+
+  @Override
+  public boolean containsAll(Collection<?> c) {
+    for (Object element : c) {
+      if (!contains(element)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  @Override
+  public boolean addAll(Collection<? extends T> c) {
+    throw new UnsupportedOperationException(
+        "modification is not supported in UnmodifiableArrayList");
+  }
+
+  @Override
+  public boolean addAll(int index, Collection<? extends T> c) {
+    throw new UnsupportedOperationException(
+        "modification is not supported in UnmodifiableArrayList");
+  }
+
+  @Override
+  public boolean removeAll(Collection<?> c) {
+    throw new UnsupportedOperationException(
+        "modification is not supported in UnmodifiableArrayList");
+  }
+
+  @Override
+  public boolean retainAll(Collection<?> c) {
+    throw new UnsupportedOperationException(
+        "modification is not supported in UnmodifiableArrayList");
+  }
+
+  @Override
+  public void clear() {
+    throw new UnsupportedOperationException(
+        "modification is not supported in UnmodifiableArrayList");
+  }
+
+  @Override
+  public T get(int index) {
+    return mElements[index];
+  }
+
+  @Override
+  public T set(int index, T element) {
+    throw new UnsupportedOperationException(
+        "modification is not supported in UnmodifiableArrayList");
+  }
+
+  @Override
+  public int indexOf(Object o) {
+    if (o == null) {
+      for (int i = 0; i < mElements.length; i++) {
+        if (mElements[i] == null) {
+          return i;
+        }
+      }
+    } else {
+      for (int i = 0; i < mElements.length; i++) {
+        if (o.equals(mElements[i])) {
+          return i;
+        }
+      }
+    }
+    return -1;
+  }
+
+  @Override
+  public int lastIndexOf(Object o) {
+    if (o == null) {
+      for (int i = mElements.length - 1; i >= 0; i--) {
+        if (mElements[i] == null) {
+          return i;
+        }
+      }
+    } else {
+      for (int i = mElements.length - 1; i >= 0; i--) {
+        if (o.equals(mElements[i])) {
+          return i;
+        }
+      }
+    }
+    return -1;
+  }
+
+  @Override
+  public ListIterator<T> listIterator() {
+    return new ListItr(0);
+  }
+
+  @Override
+  public ListIterator<T> listIterator(int index) {
+    if (index < 0 || index > mElements.length) {
+      throw new IndexOutOfBoundsException("Index: " + index);
+    }
+    return new ListItr(index);
+  }
+
+  @Override
+  public List<T> subList(int fromIndex, int toIndex) {
+    throw new UnsupportedOperationException(
+        "subList is not supported in UnmodifiableArrayList, use the underlying array instead");
+  }
```

Comment:
`List.subList` returns a view over the list like a python slice. But I don't think anyone need that at the moment.